benchmarks/conda_lock_plugin.py

```diff
@@ -0,0 +1,48 @@
+"""
+ASV plug-in providing an alternative ``Environment`` subclass, which uses Nox
+for environment management.
+
+"""
+from asv.console import log
+from asv.plugins.conda import Conda, _find_conda
+
+class CondaLock(Conda):
+    """
+    Create the environment based on a **version-controlled** lockfile.
+
+    Creating the environment instance is deferred until ``install_project`` time,
+    when the commit hash etc is known and we can access the lock file.
+    The environment is then overwritten by the specification provided at the
+    ``config.conda_lockfile`` path.  ``conda.conda_lockfile`` must point to
+    an @EXPLICIT conda manifest, e.g. the output of either the ``conda-lock`` tool,
+    or ``conda list --explicit``.
+    """
+    tool_name = "conda-lock"
+
+    def __init__(self, conf, python, requirements):
+        self._lockfile_path = conf.conda_lockfile
+        super().__init__(conf, python, requirements)
+
+    def _uninstall_project(self):
+        if self._get_installed_commit_hash():
+            # we can only run the uninstall command if an environment has already
+            # been made before, otherwise there is no python to use to uninstall
+            super()._uninstall_project()
+            # TODO: we probably want to conda uninstall all the packages too
+            #       something like:
+            #       conda list --no-pip | sed /^#/d | cut -f 1 -d " " | xargs conda uninstall
```

Comment:
> I assume you're thinking here about the small risk that a dependency might be removed between commits, perhaps if it were causing a problem, but it would not be removed from the ASV environment? Or does installing from a lock file mean that existing packages are removed?

Yes, exactly.

As it stands, this is how the environment management looks for running several commits, lets call them A->B, one after the other, each contains an updated lockfile e.b. B.lock has different packages than A.lock:

1. asv creates a new, empty conda environment `.asv/env/hash`, let's call this $ENV
2. asv checks out A.
3. asv does a `conda install --file A.lock` in $ENV.  A.lock is an explicit conda requirements file so conda doesn't resolve anything, it just installs all the listed pacakges into $ENV.
4. asv runs the benchmark for A and saves the output.
5. asv checks out B.
6. asv does a `conda install --file B.lock` in $ENV. **What happens to the packages in $ENV?** I think that if they have the same name, they are just overwritten.  Packages that were in A.lock but not in B.lock will not be removed, they just sit there.  This shouldn't be a problem, but it could be.  But I think if this was to cause and error, this should be picked up in testing, not in benchmarking.  What I'm suggesting is that we could be explicit here and clean down the environment before we do the next run.
7. asv runs benchmark for B and saves output.

I think you're right, if benchmarking could be improved to use the nox environment creation logic, this problem may go away.

Reply:
Thanks, glad I'm on the same page as you.